modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/direct/TransportPutDatabaseConfigurationAction.java

```diff
@@ -0,0 +1,198 @@
+            return Tuple.tuple(task.execute(clusterState), null);
+        }
+
+        @Override
+        public void taskSucceeded(UpdateDatabaseConfigurationTask task, Void unused) {
+            logger.trace("Updated cluster state for creation-or-update of database configuration [{}]", task.database.id());
+            task.listener.onResponse(AcknowledgedResponse.TRUE);
+        }
+    };
+
+    private final MasterServiceTaskQueue<UpdateDatabaseConfigurationTask> updateDatabaseConfigurationTaskQueue;
+
+    @Inject
+    public TransportPutDatabaseConfigurationAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        ActionFilters actionFilters,
+        IndexNameExpressionResolver indexNameExpressionResolver
+    ) {
+        super(
+            PutDatabaseConfigurationAction.NAME,
+            transportService,
+            clusterService,
+            threadPool,
+            actionFilters,
+            Request::new,
+            indexNameExpressionResolver,
+            AcknowledgedResponse::readFrom,
+            EsExecutors.DIRECT_EXECUTOR_SERVICE
+        );
+        this.updateDatabaseConfigurationTaskQueue = clusterService.createTaskQueue(
+            "update-geoip-database-configuration-state-update",
+            Priority.NORMAL,
+            UPDATE_TASK_EXECUTOR
+        );
+    }
+
+    @Override
+    protected void masterOperation(Task task, Request request, ClusterState state, ActionListener<AcknowledgedResponse> listener) {
+        final String id = request.getDatabase().id();
+
+        {
+            IngestGeoIpMetadata geoIpMeta = state.metadata().custom(IngestGeoIpMetadata.TYPE, IngestGeoIpMetadata.EMPTY);
+
+            final DatabaseConfigurationMetadata existingDatabase = geoIpMeta.getDatabases().get(id);
+            // make the request a no-op if the databases match exactly
+            if (isNoopUpdate(existingDatabase, request.getDatabase())) {
+                listener.onResponse(AcknowledgedResponse.TRUE);
+                return;
+            }
+        }
+
+        DatabaseConfiguration.validateId(id);
+
+        updateDatabaseConfigurationTaskQueue.submitTask(
+            Strings.format("update-geoip-database-configuration-[%s]", id),
+            new UpdateDatabaseConfigurationTask(listener, request.getDatabase()),
+            null
+        );
+    }
+
+    /**
+     * Returns 'true' if the database configuration is effectually the same, and thus can be a no-op update.
+     */
+    static boolean isNoopUpdate(@Nullable DatabaseConfigurationMetadata existingDatabase, DatabaseConfiguration newDatabase) {
+        if (existingDatabase == null) {
+            return false;
+        } else {
+            return newDatabase.equals(existingDatabase.database());
+        }
+    }
+
+    private static void validatePrerequisites(DatabaseConfiguration database, ClusterState state) {
```

Comment:
Can you add a unit test for this? (I haven't double-checked that you didn't already, but since it's marked `private` I assume there isn't one)